Refactor: comment module

rockpell opened the pull request:
## 바뀐점
- comment api module 추가

## 바꾼이유
- comment service의 복잡한 의존성을 낮추기 위해

## 설명
- comment api module은 api controller와 api service들을 가지고 있다
- 각각의 api service들은 comment service를 비롯한 여러 service를 가져와서 조합하여 결과를 내보낸다.
- api controller에서는 api service 하나만을 호출하고 그에 맞는 response를 리턴한다
- 객체를 수정하는 로직은 entity의 함수로 변경을 하고 repository는 저장만 하는 방식으로 처리한다.
- 댓글 작성하는 함수 entity로 이전 및 service 제거
   - create 하는 서비스 함수는 불필요하기 때문에 orm의 save 기능을 제대로 활용하는 방향으로 수정
- comment 생성, 수정, 조회 request, response dto에서 PickType 제거

rockpell committed:
Refactor: comment 모듈 리팩토링 작업 중(백업)

rockpell committed:
Refactor: 댓글 생성 로직, api service로 이전

rockpell committed:
Test: e2e에 CommentApiModule 적용

rockpell committed:
Feat: comment entity에 updateContent 추가

- content를 변경하는 함수 추가

rockpell committed:
Refactor: comment를 수정하는 api 관련 함수 api service로 이전

- comment service에 있던 수정하는 함수를 api service로 이전
- 객체를 수정하는 로직은 entity에 있고 service는 찾거나 저장하는 역할만 수행
- api service는 controller에서 직접 호출되는 service이므로 service와 객체의 함수 조합으로 구성
- service에 있어도 될것 같은 함수이지만 아직은 어디에 둘지 잘 모르겠어서 api service로 이전하였음

rockpell committed:
Refactor: 댓글 삭제 기능, remove-comment-api.service로 이전

Skyrich2000 commented on libs/entity/src/comment/comment.entity.ts:
```diff
@@ -63,4 +63,8 @@ export class Comment extends BaseEntity {
   })
   reactionComment?: ReactionComment[];
+
+  public updateContent(content: string) {
+    this.content = content
+  }
```
지금 entity 가 libs 에 있어서 뭔가 비지니스를 담으려니까 애매하다고 느꼈었는데
당장은 어쩔수없는거같기도 하네요 나중에 도메인객체 따로 생기면 그떄 또 리펙토링하는게 나을거같네영

rockpell replied:
네 일단은 넣어두면 나중에 옮기기 쉬울테니 해두려구요

Skyrich2000 commented:
*-comment-api.service 에서 로직을 호출하고 comment.service 는 repository를 한번 감싸기만한거같은 느낌이군요 
그리고 api 별로 서비스를 따로 두어서 각 서비스별로 책임을 더 분산하니까 훨씬 가독성이 좋아지네요 (파일이 좀 많아지긴 하지만 ㅋ.ㅋ)

막상 이렇게 나누고 보니까 comment-api.service 들이 사실상 비지니스 layer 가 된거같네요 오히려 comment.service 의 역할이 이름대비 기능을 제대로 못하는거같기도하고,, 네이밍은 다시 생각해볼만한거같아요..

Skyrich2000 commented on apps/api/src/comment/comment.service.ts:
```diff
@@ -69,30 +58,19 @@ export class CommentService {
 
-  async updateContent(id: number, writerId: number, updateCommentDto: UpdateCommentRequestDto): Promise<void | never> {
-    const comment = await this.commentRepository.findOneOrFail({
+  async findByIdAndWriterIdOrFail(id: number, writerId: number): Promise<Comment> {
+    return this.commentRepository.findOneOrFail({
       id,
       writerId,
     });
```
요거 article 에도 비슷하게 있었는데 솔직히 그냥 id 만으로도 충분히 query 할 수 있으니까 굳이 writerId 를 따로 같이 안넣어도 될거같아요 
대신에 글쓴이 본인인지 확인하는 로직을 comment 에 포함하는방향이 더 나을거같아요
취향차이 일 수 있을거같긴한데, 로직에 의도가 숨겨져있는 느낌이라고 생각했슴다 어떻게 생각하십니까

rockpell replied:
그런 방식이 좋긴하죠
나중에 적용해보겠습니다
writerId 빼려다가 예외처리 관련 테스트 코드가 있을거 같아서 그냥 두긴했네요

rockpell replied:
> *-comment-api.service 에서 로직을 호출하고 comment.service 는 repository를 한번 감싸기만한거같은 느낌이군요 그리고 api 별로 서비스를 따로 두어서 각 서비스별로 책임을 더 분산하니까 훨씬 가독성이 좋아지네요 (파일이 좀 많아지긴 하지만 ㅋ.ㅋ)
> 
> 막상 이렇게 나누고 보니까 comment-api.service 들이 사실상 비지니스 layer 가 된거같네요 오히려 comment.service 의 역할이 이름대비 기능을 제대로 못하는거같기도하고,, 네이밍은 다시 생각해볼만한거같아요..

comment.service가 repository를 호출하는 persistence 어댑터의 역할을 하고 있어서 그냥 어댑터로 이름을 바꾸는게 나을거 같기도하네요
다른 모듈은 아직 정리가 안됐으니 의존성 분리 끝나면 이름만 바꾸면 될거 같습니다

Skyrich2000 replied:
네 저도 어댑터가 더 맞는이름인거같아요 리펙토링 슥슥 하면서 바꾸면 될거같습니다 👍

rockpell committed:
Merge branch 'develop' into refactor/comment-module

rockpell committed:
Refactor: comment service에서 다른 service 의존하는 코드 이전

- article-api.service의 getComment 함수에서 여러 서비스를 호출하도록 수정
- comment service에서 다른 service를 의존하고 있던 문제 수정

Skyrich2000 commented on apps/api/src/comment/comment.service.ts:
```diff
@@ -29,19 +21,12 @@ export class CommentService {
+  }> {
     const { comments, totalCount } = await this.commentRepository.findAllByArticleId(articleId, options);
-    return { comments, category, totalCount };
+    return { comments, totalCount };
```
이거 그냥 바로 return await 해도 될거같네영

rockpell replied:
오 이거 방금 커밋했는데 바로 보셨군요

rockpell committed:
Refactor: 불필요한 변수 할당 제거

rockpell committed:
Refactor: comment service 디렉토리 추가

rockpell committed:
Chore: 불필요한 주석 제거

Skyrich2000 commented:
이미 폴더자체가 여러개를 담고있다는거니까 복수형말고 단수형으로 해도 괜찮지 않을까염.. ex) 그냥 service

Skyrich2000 commented:
module 별로 폴더를 구분하는게 나을거같기두하네염 흠....

Skyrich2000 commented:
CreateCommentApiService

Skyrich2000 commented:
CreateCommentApiService vs CommentCreateApiService
UpdateCommentApiService vs CommentUpdateApiService
RemoveCommentApiService vs CommentRemoveApiService

prefix 가 Comment 로 시작하니까 일괄성있어서 좋은데 단어 어순은 안맞는거같아서 고민이군염 흠...
좀더 복잡한 Service 가 나오면 오히려 어순이 맞게 하는게 나을거같긴하네용

rockpell commented:
> CreateCommentApiService vs CommentCreateApiService UpdateCommentApiService vs CommentUpdateApiService RemoveCommentApiService vs CommentRemoveApiService
> 
> prefix 가 Comment 로 시작하니까 일괄성있어서 좋은데 단어 어순은 안맞는거같아서 고민이군염 흠... 좀더 복잡한 Service 가 나오면 오히려 어순이 맞게 하는게 나을거같긴하네용

저는 통일성보다는 어순이 맞는데 더 좋다고 생각합니다.
참고로 만들면서 배우는 클린아키텍처 책에서는 지금 같은 형태의 네이밍으로 클래스를 만들긴합니다

rockpell commented:
> 이미 폴더자체가 여러개를 담고있다는거니까 복수형말고 단수형으로 해도 괜찮지 않을까염.. ex) 그냥 service

단수로 해도 되긴하는데 repositories를 복수형으로 해놔서 일단 복수형으로 했습니다
나중에 전부 단수형으로 바꿔도 좋을거 같네요

rockpell committed:
Refactor: 댓글 가져오는 api, comment 디렉토리로 이전

- article comment에 댓글 가져오는 controller가 있어서 의존성이 이상하게 흐르는 문제 수정
- get-comment api를 위한 별도의 컨트롤러 추가

rockpell committed:
Test: article e2e에 있던 get-commnet api 테스트 이전

- 댓글 가져오는 api를 comment 모듈로 이전하면서 테스트도 그에 맞게 수정

rockpell committed:
Refacotr: comment create, update, request 객체 수정

- pickType으로 필요한 프로퍼티만 뽑아서 사용하는 방식을 사용하고 있었으나 직관성이 부족해서 수정

rockpell committed:
Fix: 댓글 생성 api, swagger 객체 잘못 들어간거 문제

- swagger의 response 자리에 request 객체를 넣어서 문서가 이상하게 만들어지고 있던 문제 수정

rockpell committed:
Refactor: CommentResponseDto PickType 제거

rockpell committed:
Refactor: 댓글 작성하는 함수 entity로 이전 및 service 제거

- create 하는 서비스 함수는 불필요하기 때문에 orm의 save 기능을 제대로 활용하는 방향으로 수정

Skyrich2000 commented:
어자피 service 폴더에 격리되어있으니까 어순이 맞게 하는게 더 나을거같네용
그리고 나중에 repositories 도 단수형으로 바꾸는게 좋을거같아요

Skyrich2000 commented on libs/entity/src/comment/comment.entity.ts:
```diff
@@ -64,6 +64,15 @@ export class Comment extends BaseEntity {
   reactionComment?: ReactionComment[];
 
+  // TODO: 지금은 BaseEntity 때문에 함수이름이 겹쳐서 조금 이상한 이름, 추후 그냥 create로 변경 해야함
+  public static createComment(content: string, articleId: number, writerId: number) {
```
혹시 인자를 객체로 묶어서 받으시는건 어떠십니까 취향 차이이긴한데, 가져다 쓸때 객체 key 값으로 인자를 구분할 수 있어서 좀더 좋은거같았어요
뭐 어짜피 ide 에서 parameterhint 가 뜨긴하지만,,,,

rockpell replied:
음 확실히 ts에서는 그게 낫겠네요
코틀린처럼 해버렸네요

Skyrich2000 commented on apps/api/src/comment/get-comment-api.controller.ts:
```diff
@@ -0,0 +1,44 @@
+@ApiCookieAuth()
+@ApiUnauthorizedResponse({ description: '인증 실패' })
+@ApiTags('Comment')
+@Controller('articles')
+export class GetCommentApiController {
```
GetCommentApiController 인데 api 경로가 articles 라니... 
이정도면 기존 api 스펙을 바꾸는게 낫지 않을까 싶네요.. 흠...

Skyrich2000 replied:
기존에 article api controller 에서 comment service 를 의존하는 방식이 의존성이 이상했나요? 이거 한번 sync 맞추면 좋을거같아요

Skyrich2000 replied:
이거 지금 스펙대로 진행하고, 나중에 프론트 마이그레이션 되면 삭제하는걸로 얘기했슴당

Yaminyam replied:
comment가 article에 종속되어있는 형태기때문에 엔드포인트자체는 괜찮다고 생각이되는데 도메인 설계관점으로 보면 또 이상하긴 하네요

Skyrich2000 replied:
restful 하게 하려고 한건데 이마저도 이상한거같아요 ㅋㅋ
이쯤되면 graphql 할까싶을정도...

rockpell committed:
Refactor: Comment 생성 함수, 파라미터 객체로 변경

Skyrich2000 commented:
개인적으로 폴더별로 index.ts 를 만들어서 import 할떄 묶어서 import 하면 좋을거같아요 
article 이거 안되어있어서 저도 지금 리펙토링 하는중 ㅋㅋㅋ

Skyrich2000 commented on libs/entity/src/comment/comment.entity.ts:
```diff
@@ -2,17 +2,17 @@ import { Article } from '@app/entity/article/article.entity';
-  OneToMany,
-  PrimaryGeneratedColumn,
-  UpdateDateColumn,
+BaseEntity,
+Column,
+CreateDateColumn,
+DeleteDateColumn,
+Entity,
+Index,
+JoinColumn,
+ManyToOne,
+OneToMany,
+PrimaryGeneratedColumn,
+UpdateDateColumn
```
이거 lint 가 좀 수상한거같아요..!

Skyrich2000 commented on apps/api/src/comment/services/create-comment-api.service.ts:
```diff
@@ -20,7 +20,11 @@ export class CreateCommentApiService {
 
-    const comment = Comment.createComment(createCommentDto.content, createCommentDto.articleId, writer.id);
+    const comment = Comment.createComment({
+      content: createCommentDto.content,
+      articleId: createCommentDto.articleId,
```
개인적으로는 service 레이어까지 DTO 가 최대한 안넘오는게 좋은거같은데 어떻게 생각하시나유

rockpell replied:
service 함수의 모든 파라미터는 command 객체로 만들어서 받는게 제일 좋긴하죠

rockpell replied:
일단 급한대로 그냥 리터럴 객체로 받아야겠네요

Skyrich2000 replied:
command 객체라는걸 따로 만드는군요 이건 몰랐넹

Skyrich2000 replied:
그냥 함수 인자로 풀어서 넣어줘도 될거같긴해요 
아까 createComment 는 객체로 한번 감싸달라고 하긴했는데.. 가로로 너무 길어지지만 않으면 괜찮다고 생각함당..........
뭔가 이래라저래라하는거같으니 이건 채림님 취향에 맡기겠슴당

rockpell replied:
일단은 이렇게 두고 나중에 필요하면 command 객체 만드는걸로 하시죠

rockpell committed:
Fix: prettier 오작동으로 인한 재적용

rockpell committed:
Refactor: 댓글 생성 service 함수, 파라미터 타입 변경

- dto를 서비스에서 사용하지 않기 위해 리티럴 객체로 받는 방식으로 변경

Skyrich2000 commented on apps/api/src/comment/comment-api.controller.ts:
```diff
@@ -20,23 +22,31 @@ import { CommentResponseDto } from './dto/response/comment-response.dto';
-    @Body() createCommentDto: CreateCommentRequestDto,
-  ): Promise<CommentResponseDto | never> {
-    const comment = await this.commentService.create(writer, createCommentDto);
+    @Body() request: CreateCommentRequestDto,
```
```suggestion
    @Body() { content, articleId }: CreateCommentRequestDto,
```

Skyrich2000 commented on apps/api/src/comment/comment-api.controller.ts:
```diff
@@ -55,17 +65,17 @@ export class CommentController {
-    @Body() updateCommentDto: UpdateCommentRequestDto,
-  ): Promise<void | never> {
-    return this.commentService.updateContent(id, writer.id, updateCommentDto);
+    @Body() request: UpdateCommentRequestDto,
```
```suggestion
    @Body() { content }: UpdateCommentRequestDto,
```

rockpell committed:
Merge branch 'develop' into refactor/comment-module

Skyrich2000 reviewed (approved):
이거 슬슬 머지 가시쥬

rockpell committed:
Merge branch 'develop' into refactor/comment-module
